File: src/in/benjamm/pms/DataModel/FolderScanning/FolderScanOperation.java
package in.benjamm.pms.DataModel.FolderScanning;

import in.benjamm.pms.DataModel.Model.Folder;
import in.benjamm.pms.DataModel.Model.MediaItem;
import in.benjamm.pms.DataModel.Model.Song;
import org.jaudiotagger.audio.AudioFile;
import org.jaudiotagger.audio.AudioFileIO;
import org.jaudiotagger.audio.exceptions.CannotReadException;
import org.jaudiotagger.audio.exceptions.InvalidAudioFrameException;
import org.jaudiotagger.audio.exceptions.ReadOnlyFileException;
import org.jaudiotagger.tag.TagException;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;

import static in.benjamm.pms.DataModel.Singletons.Log.*;

/**
 * Created with IntelliJ IDEA.
 * User: bbaron
 * Date: 6/18/12
 * Time: 5:12 PM
 * To change this template use File | Settings | File Templates.
 */
public class FolderScanOperation extends ScanOperation
{
    private String _folderPath;
    public String getFolderPath() { return _folderPath; }

    private String[] _validExtensions = {"mp3", "aac", "m4a", "mp4", "flac", "wv", "mpc", "ogg"};
    private List<String> _validExtensionsList = Arrays.asList(_validExtensions);

    public FolderScanOperation(String folderPath, int secondsDelay)
    {
        super(secondsDelay);
        _folderPath = folderPath;
    }

    public void start()
    {
        processFolder(getFolderPath());
    }

    /**
     * Scan a folder for file changes
     */

    /*
     * Public methods
     */

    public void processFolder(int folderId)
    {
        Folder folder = new Folder(folderId);
        processFolder(folder.getFolderPath());
    }

    public void processFolder(String folderPath)
    {
        if (isRestart())
            return;

        File topFile = new File(folderPath);
        if (topFile.isDirectory())
        {
            // Retreive the folder object for this folder path
            final Folder topFolder = new Folder(topFile.getAbsolutePath());
            log2Out(TEST, "scanning " + topFolder.getFolderName() + "  id: " + topFolder.getFolderId());

            // Start recursively scanning the subfolders and files
            // Note: These won't be in any particular order
            for (final File subFile : topFile.listFiles())
            {
                if (isRestart())
                    return;

                if (subFile.isDirectory())
                {
                    if (!subFile.getName().startsWith(".AppleDouble"))
                    {
                        // This is a folder, so scan it
                        Folder folder = new Folder(subFile.getAbsolutePath());
                        if (folder.getFolderId() == null)
                        {
                            // This folder isn't in the database, so add it
                            folder.addToDatabase();
                        }
                        processFolder(subFile.getAbsolutePath());
                    }
                }
                else
                {
                    submitTask(new Runnable() {
                        public void run() {
                            processFile(subFile, topFolder.getFolderId());
                        }
                    });
                }
            }
        }
    }

    public void processFile(File file, int folderId)
    {
        if (isRestart())
            return;

        if (file.getName().startsWith(".DS_Store"))
            return;

        // Make sure the extension is valid
        String fileName = file.getName();
        int index = fileName.lastIndexOf(".");
        String extension = fileName.substring(index+1, fileName.length());
        if (!_validExtensionsList.contains(extension.toLowerCase()))
            return;

        //log2Out(TEST, "processing file " + file.getName());

		if (MediaItem.fileNeedsUpdating(file))
		{
			AudioFile f = null;
			try {
				f = AudioFileIO.read(file);
			} catch (CannotReadException e) {
                log2Out(TEST, "Can't read file " + file.getName());
			} catch (IOException e) {
                log2Out(TEST, "Can't read file " + file.getName());
			} catch (TagException e) {
                log2Out(TEST, "Can't read file " + file.getName());
			} catch (ReadOnlyFileException e) {
                log2Out(TEST, "Can't read file " + file.getName());
			} catch (InvalidAudioFrameException e) {
                log2Out(TEST, "Can't read file " + file.getName());
			}

			if (f == null)
            {
                // This is either another media type, a random file, or just not supported by jAudioTagger
            }
            else
			{
				// This is a song, process it
				Song song = new Song(f, folderId);
				song.updateDatabase();
			}
		}
    }

    public int hashCode()
    {
        return getFolderPath().hashCode();
    }

    public boolean equals(Object obj)
    {
        if (obj == null)
            return false;
        if (obj == this)
            return true;
        if (obj.getClass() != getClass())
            return false;

        FolderScanOperation otherOperation = (FolderScanOperation)obj;
        return getFolderPath().equals(otherOperation.getFolderPath());
    }
}

File: pms/DataModel/FolderScanning/FolderScanOperation.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using WaveBox.DataModel.Model;
using WaveBox.DataModel.Singletons;
using System.IO;
using System.ComponentModel;
using System.Threading.Tasks;
using System.Diagnostics;
using TagLib;

namespace WaveBox.DataModel.FolderScanning
{
	class FolderScanOperation : ScanOperation
	{
		private string _folderPath;
		public string FolderPath
		{
			get
			{
				return _folderPath;
			}
		}

		private string[] _validExtensions = { "mp3", "m4a", "mp4", "flac", "wv", "mpc", "ogg", "wma" };
		private List<string> _validExtensionsList;

		public FolderScanOperation(string folderPath, int secondsDelay) : base(secondsDelay)
		{
			_validExtensionsList = new List<string>(_validExtensions);
			_folderPath = folderPath;
		}

		public override void Start()
		{
			processFolder(FolderPath);
		}

		public void processFolder(int folderId)
		{
			var folder = new Folder(folderId);
			processFolder(folder.FolderPath);
		}

		public void processFolder(string folderPath)
		{
			if (ShouldRestart)
			{
				return;
			}

			FileInfo topFile;

			try
			{
				topFile = new FileInfo(folderPath);

				if (topFile.Directory.Exists == false)
				{
					return;
				}

				Folder topFolder;

				// if the file is a directory
				if ((topFile.Attributes & FileAttributes.Directory) == FileAttributes.Directory)
				{
					topFolder = new Folder(topFile.FullName);
					//Console.WriteLine("scanning " + topFolder.FolderName + "  id: " + topFolder.FolderId);

					foreach (var subfolder in Directory.GetDirectories(topFile.FullName))
					{
						if (!(subfolder.Contains(".AppleDouble")))
						{
							var folder = new Folder(subfolder, true);

							// if the folder isn't already in the database, add it.
							if (folder.FolderId == 0)
							{
								folder.addToDatabase(false);
							}
							processFolder(subfolder);
						}
					}

					//Parallel.ForEach(Directory.GetDirectories(topFile.FullName), currentFile =>
					//    {
					//        if (!(currentFile.Contains(".AppleDouble")))
					//        {
					//            var folder = new Folder(currentFile);

					//            // if the folder isn't already in the database, add it.
					//            if (folder.FolderId == 0)
					//            {
					//                folder.addToDatabase(false);
					//            }
					//            processFolder(currentFile);
					//        }
					//    });

					foreach (var subfile in Directory.GetFiles(topFile.FullName))
					{
						// if the subfile is a directory...
						processFile(new FileInfo(subfile), topFolder.FolderId);
						//Console.WriteLine(subfile);
					}

					//Parallel.ForEach(Directory.GetFiles(topFile.FullName), currentFile =>
					//    {
					//        processFile(new FileInfo(currentFile), topFolder.FolderId);
					//    });



				}
			}

			catch (FileNotFoundException e)
			{
				Console.WriteLine("\t" + "[FOLDERSCAN] " + folderPath + ": Directory does not exist. " + e.InnerException);
			}

			catch (DirectoryNotFoundException e)
			{
				Console.WriteLine("\t" + "[FOLDERSCAN] " + folderPath + ": Directory does not exist. " + e.InnerException);
			}
			catch (IOException e)
			{
				Console.WriteLine("\t" + "[FOLDERSCAN] " + folderPath + ": " + e.Message);
			}
			catch (Exception e)
			{
				Console.WriteLine("\t" + "[FOLDERSCAN] " + "Error checking to see if the file was a directory: " + e.ToString());
			}
		}

			public void processFile(FileInfo file, int folderId)
			{
				if (ShouldRestart)
				{
					return;
				}

				// make sure the extension is valid
				if (!_validExtensionsList.Contains(file.Extension.Substring(1).ToLower()))
				{
					return;
				}


				if (MediaItem.fileNeedsUpdating(file))
				{
					Console.WriteLine("[FOLDERSCAN] " + "File needs updating: " + file.Name);
					var sw = new Stopwatch();
					sw.Start();
					TagLib.File f = null;
					try
					{
						f = TagLib.File.Create(file.FullName);
					}

					catch (TagLib.CorruptFileException e)
					{
						e.ToString();
						Console.WriteLine("[FOLDERSCAN] " + file.Name + " has a corrupt tag and will not be inserted.");
						return;
					}

					catch (Exception e)
					{
						Console.WriteLine("[FOLDERSCAN] " + "Error processing file: " + e.ToString());
					}
					sw.Stop();
					//Console.WriteLine("Get tag: {0} ms", sw.ElapsedMilliseconds);

					sw.Reset();

					if (f == null)
					{
						// Must be something not supported by TagLib-Sharp
					}

					else
					{
						// It's a song!  Do yo thang.
						sw.Start();
						var song = new Song(file, folderId);
						sw.Stop();
						//Console.WriteLine("Create new song object: {0} ms", sw.ElapsedMilliseconds);

						sw.Reset();
						sw.Start();
						song.updateDatabase();
						sw.Stop();
						//Console.WriteLine("Update database: {0} ms", sw.ElapsedMilliseconds);
					}
				}
			}

			public override string ScanType()
			{
				return String.Format("FolderScanOperation:{0}", FolderPath);
			}

		}
	}


